src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java

```diff
@@ -189,43 +187,54 @@ public int getOutstandingRequests() {
   }
 
   @Override
-  public <T> ListenableFuture<T> send(final Request<T> request) {
-    if (!tryIncrementPending()) {
+  public <T> CompletableFuture<T> send(final Request<T> request) {
 
+    if (!tryIncrementPending()) {
       // Do the disconnect check in here instead of outside
       // to get better performance in the happy case.
       String disconnectReason = this.disconnectReason.get();
       if (disconnectReason != null) {
-        MemcacheClosedException exception = new MemcacheClosedException(disconnectReason);
-        return onExecutor(Futures.<T>immediateFailedFuture(exception));
+        return newExceptionallyCompletedFuture(new MemcacheClosedException(disconnectReason));
```

Comment:
why remove onExecutor? Is the default for Guava futures to run callbacks on the same thread, while Java 8 always run on a common FJP as default? If so, then I am ok with this.

My main concern is users running callbacks on netty threads

Reply:
Yeah, they are run on the default fork-join pool. Most likely (but not guaranteed) on the same thread, since it is a bit biased when deploying next task. 

The idea was that it is simpler without an additional pool.

Reply:
This depends on the method (e.g. `thenFoo` vs `thenFooAsync`) and whether the future is already completed or not.

In this case, the future is already complete so `thenFoo(cb)` will invoke the callback on the caller thread and `thenFooAsync(cb)` will [typically](http://hg.openjdk.java.net/jdk9/jdk9/jdk/file/tip/src/java.base/share/classes/java/util/concurrent/CompletableFuture.java#l429) invoke it on the common FJP. I think that's fine for all immediately completed futures (i.e. errors).

For futures that are not yet completed, `thenFoo(cb)` will cause the callback to be invoked on the _completing_ thread. So by keeping the `onExecutor` wrapping of the actual request future on [line 215](https://github.com/spotify/folsom/pull/80/files#diff-1b1a61f1bf891fde7e8125494305e141R215) we avoid user callbacks being invoked on netty IO threads.